sandbox/cgroup/inotify/inotify_linux.go

```diff
@@ -0,0 +1,315 @@
+	}
+	w.isClosed = true
+
+	// Send "quit" message to the reader goroutine
+	w.done <- true
+	for path := range w.watches {
+		w.RemoveWatch(path)
+	}
+
+	return nil
+}
+
+// AddWatch adds path to the watched file set.
+// The flags are interpreted as described in inotify_add_watch(2).
+func (w *Watcher) AddWatch(path string, flags uint32) error {
+	if w.isClosed {
+		return errors.New("inotify instance already closed")
+	}
+
+	watchEntry, found := w.watches[path]
+	if found {
+		watchEntry.flags |= flags
+		flags |= syscall.IN_MASK_ADD
+	}
+
+	w.mu.Lock() // synchronize with readEvents goroutine
+
+	wd, err := syscall.InotifyAddWatch(w.fd, path, flags)
+	if err != nil {
+		w.mu.Unlock()
+		return &os.PathError{
+			Op:   "inotify_add_watch",
+			Path: path,
+			Err:  err,
+		}
+	}
+
+	if !found {
+		w.watches[path] = &watch{wd: uint32(wd), flags: flags}
+		w.paths[wd] = path
+	}
+	w.mu.Unlock()
+	return nil
+}
+
+// Watch adds path to the watched file set, watching all events.
+func (w *Watcher) Watch(path string) error {
+	return w.AddWatch(path, InAllEvents)
+}
+
+// RemoveWatch removes path from the watched file set.
+func (w *Watcher) RemoveWatch(path string) error {
+	watch, ok := w.watches[path]
+	if !ok {
+		return fmt.Errorf("can't remove non-existent inotify watch for: %s", path)
+	}
+	success, errno := syscall.InotifyRmWatch(w.fd, watch.wd)
+	if success == -1 {
+		// when file descriptor or watch descriptor not found, InotifyRmWatch syscall return EINVAL error
+		// if return error, it may lead this path remain in watches and paths map, and no other event can trigger remove action.
+		if errno != syscall.EINVAL {
+			return os.NewSyscallError("inotify_rm_watch", errno)
+		}
+	}
+	delete(w.watches, path)
+	// Locking here to protect the read from paths in readEvents.
+	w.mu.Lock()
+	delete(w.paths, int(watch.wd))
+	w.mu.Unlock()
+	return nil
+}
+
+// readEvents reads from the inotify file descriptor, converts the
+// received events into Event objects and sends them via the Event channel
+func (w *Watcher) readEvents() {
+	var buf [syscall.SizeofInotifyEvent * 4096]byte
+
+	for {
+		n, err := syscall.Read(w.fd, buf[:])
+		// See if there is a message on the "done" channel
+		var done bool
+		select {
+		case done = <-w.done:
+		default:
+		}
+
+		// If EOF or a "done" message is received
+		if n == 0 || done {
+			// The syscall.Close can be slow.  Close
+			// w.Event first.
+			close(w.Event)
+			err := syscall.Close(w.fd)
+			if err != nil {
+				w.Error <- os.NewSyscallError("close", err)
+			}
+			close(w.Error)
+			return
+		}
+		if n < 0 {
+			w.Error <- os.NewSyscallError("read", err)
+			continue
+		}
+		if n < syscall.SizeofInotifyEvent {
+			w.Error <- errors.New("inotify: short read in readEvents()")
+			continue
+		}
+
+		var offset uint32
+		// We don't know how many events we just read into the buffer
+		// While the offset points to at least one whole event...
+		for offset <= uint32(n-syscall.SizeofInotifyEvent) {
```

Comment:
Nitpick: I find this easier to read:
```suggestion
		for offset + syscall.SizeofInotifyEvent <= n {
```
(add the cast if/where needed)

Reply:
Honestly, I prefer to avoid touching that part, because it is a low-level code... You know... "if it works"...